### ISSUE TITLE
Replace functions from `utils` with upstream functions

### ISSUE THREAD

**Issue by ranjithshegde** [neovim/nvim-lspconfig#2079]
### Language server

N/A

### Requested feature

1. `path`
2. `search_ancestors`
3. `root_pattern`
4. `find_git_ancestors`
5. `find_mercurial_ancestors`
6. `find_node_modules_ancestors`
7. `find_package_json_ancestors`

All of the above can be replaced with relevant functions from `vim.fs` module.

## Proposal
1. Soft deprecate it after the release of `nvim 0.8` + stabilization period. 
2. Internally redirect the calls to the relevant calls from `vim.fs` module
3. replaces the functions inside `server_configurations/[server].lua` wherever relevant
4. Provide examples and comprehensive documentation

### Other clients which have this feature

N/A

**Comment by justinmk** [comment]
@ranjithshegde this would still be a welcome change, if you are keen

**Comment by glepnir** [comment]
there are some external compatibility issues. those functions not only used  in our repo.. so  I am a little worried about the issues that may arise from the changes.

**Comment by justinmk** [comment]
> those functions not only used in our repo.

Where are they used? lspconfig was never intended to be a platform. It's intended to be data-only. And we need to move towards that. This is the way we will be able to improve quality and improve maintainability of these configs.

**Comment by glepnir** [comment]
I knwo and agree your thought..but it used in some other plugins or user config .just a quick search you will see 
https://github.com/search?q=lspconfig.util&type=code&p=3

**Comment by justinmk** [comment]
We don't necessarily need to remove `lspconfig.util.root_pattern`, we can start by updating their implementations to use vim.fs and minimizing code where possible. And where possible, show a deprecation message that explains how to use `vim.fs` directly instead of e.g. `lspconfig.util.root_pattern`. And update all of the examples in lspconfig docs to use `vim.fs` directly, instead of `lspconfig.util`.

**Comment by justinmk** [comment]
For example, `vim.fs.root()` exists now.

**Comment by glepnir** [comment]
yes.. just worried that some util functions  like util.pathjoin etc  abused and cause incompatibility issues. So I have not refactored here.

**Comment by dundargoc** [comment]
Since we don't actually test/run any of the lua configs in CI to my knowledge, the best we can hope for is to make a string search of changed lines on PRs and complain if the deprecated function is used.

**Comment by justinmk** [comment]
The "Proposal" in the main description of this issue is updated and correct. To be clear:

- we don't need to remove anything on `util`, its interface can stay frozen for back-compat.
- *internally*, the implementation of `util` should *shrink* as we update it to use stdlib features of the minimum-supported Nvim version.
- as we bump the minimum required Nvim, we will add lint rules to the lspconfig CI which prevent configs from using `util` functions if the minimum supported Nvim (currently 0.9) has alternatives.

**Comment by justinmk** [comment]
# Status

related work: 

- https://github.com/neovim/nvim-lspconfig/pull/3450

**Comment by dundargoc** [comment]
I'm not really a fan of the suggested deprecation strategy. It creates extra CI work and makes the process harder because we now need to have bookkeeping of which util functions we use ourselves and which util functions are there for public use. What was so wrong about the previous deprecation strategy that we need to invent a new one on the spot?

**Comment by justinmk** [comment]
Ok I see what you mean from the discussion in https://github.com/neovim/nvim-lspconfig/pull/3460

**Comment by justinmk** [comment]
https://github.com/neovim/nvim-lspconfig/pull/3460 is a good example of why removing `util.dirname` gains almost nothing. Keeping the skeleton around is low cost. Because we'll just freeze it once everything is upstreamed. And meanwhile, shrinking the actual implementation gives us a clear picture of which logic still doesn't have an upstream equivalent. The mere presence of `util.dirname` doesn't harm anything, though, yes, we will need to update something like `.github/ci/run_sanitizer.sh` to prevent future uses of `util.dirname` in configs.

As we hollow-out the guts of `util`, it will become almost empty. We can later decide whether we want to delete it entirely, and that can be done as one big "breaking" step instead of many incremental "breaking" steps.

**Comment by dundargoc** [comment]
> Which part of the above plan, specifically, do you want to change, and how, specifically?

The proposed strategy is alright I guess, it's just more safe/careful than my personal preference. And I like having one deprecation strategy over all our repos if possible to reduce the mental overhead of keeping track of different strategies.

But I can work with the strategy you outlined, it's not a dealbreaker and we can make progress like you said.

**Comment by justinmk** [comment]
Only thing remaining is `root_pattern` , `insert_package_json`, and `strip_archive_subpath`

https://github.com/neovim/nvim-lspconfig/issues/3768
